### DISCUSSION TITLE
Data Validation Issue When Modifying the Address Bar in Next.js 14

### DISCUSSION THREAD

**Discussion by Sevansi** [TanStack/query#7288] [Q&A] [answered]
**Context:** Using Next.js version 14.2.1 along with `tanstack/react-query` version 5.29.2. The application features a mechanism where the category ID is dynamically added to the URL when operations are performed in a modal window (for instance, updating a category). This facilitates maintaining context across page reloads and enhances user experience by providing smoother navigation.

**Problem:** When a user opens the modal window and performs an update, the category ID in the URL changes dynamically. After sending data to the server, two queries are invalidated through `react-query`. However, one of the queries (`getCategoriesById-21`) fails to complete and remains in a loading state. This issue occurs when the category ID is removed from the URL, likely affecting the query's state update process.

**Reproducing the Problem:**
1. Open the sandbox environment.
2. Open developer tools and go to the devtools tab for `tanstack-query`.
3. Open the modal window and click 'click me'.
4. Observe how the query `getCategoriesById-21` gets stuck in the loading state after the `id` is removed from the address bar.

LINK: [sanbox](https://codesandbox.io/p/github/Sevansi/tanstack-query-test/test?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clv2a8x2z0006356h3tp740xt%2522%252C%2522sizes%2522%253A%255B60%252C40%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clv2a8x2y0002356hwko8jy8n%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clv2a8x2y0004356hkp4uegfx%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clv2a8x2y0005356h4c13muja%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clv2a8x2y0002356hwko8jy8n%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv2a8x2y0001356hcqohhghq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clv2a8x2y0002356hwko8jy8n%2522%252C%2522activeTabId%2522%253A%2522clv2a8x2y0001356hcqohhghq%2522%257D%252C%2522clv2a8x2y0005356h4c13muja%2522%253A%257B%2522id%2522%253A%2522clv2a8x2y0005356h4c13muja%2522%252C%2522activeTabId%2522%253A%2522clv2ac7td003o356h956hfdwc%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clv2ac7td003o356h956hfdwc%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clv2a8x2y0004356hkp4uegfx%2522%253A%257B%2522id%2522%253A%2522clv2a8x2y0004356hkp4uegfx%2522%252C%2522activeTabId%2522%253A%2522clv2a8x2y0003356hrusgkucb%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv2a8x2y0003356hrusgkucb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clv2adgz8001idki77cdd5ff8%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522CSB_RUN_OUTSIDE_CONTAINER%253D1%2520devcontainer%2520templates%2520apply%2520--template-id%2520%255C%2522ghcr.io%252Fdevcontainers%252Ftemplates%252Ftypescript-node%255C%2522%2520--template-args%2520%27%257B%257D%27%2520--features%2520%27%255B%255D%27%2522%252C%2522id%2522%253A%2522clv2a95sa003o356hcol0u2s1%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clv2a97al007e356hnqdfqdv9%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by TkDodo** [comment]
not exactly sure what happens here, but if you keep the mutation pending while the invalidations happen, you don't run into this:

```
    onSuccess: async () => {
      await Promise.all([
        queryClient.invalidateQueries({ queryKey: ["getAllCategories"] }),
        queryClient.invalidateQueries({
          queryKey: [`getCategoriesById-${21}`],
        }),
      ]);
```

**Reply by Sevansi** [reply]
Yes, it works, I've already tested it, but there are a few nuances. For instance, when the component mounts, if there are requests that coincide with changes to the address bar, it gets stuck on loading. I'm still trying to understand the reason behind this behavior.

**Reply by TkDodo** [reply]
> I'm still trying to understand the reason behind this behavior.

yeah it's weird, I don't really know if this is next specific or not. Could you try reproducing it with a client side router, like `react-router` ?

**Reply by Sevansi** [reply]
I'm actually confused; I tried to replicate it on plain React and you know what, there's no such issue there! Here's the link [LINK](https://codesandbox.io/p/github/Sevansi/tanstack-query-test-react/main). It seems that Next.js somehow blocks the stack when updating the address bar, which prevents tanstack-query from understanding whether it has finished or not. I'm not sure what the cause is, but the standard Next.js tools like useRouter are clearly doing something.

**Reply by Sevansi** [reply]
I continued to develop my project and encountered a recurring issue. I have a page for creating items and another page where you can view all items. Logically, after adding an item, it should redirect to the page where you can view all items. However, any modification in the address bar during data validation breaks the process, and it remains stuck in loading. Initially, I thought the bug only occurred when appending parameters or changing them via replace, but it turns out that any interaction with the route in Next.js disrupts the validation by tanstack-query. I'm not sure whether the issue lies with Next.js or tanstack-query.

Regarding the suggestion to use Promise.all, it could work as it ensures we wait an additional 0.2-0.5 seconds to be certain all requests have been invalidated before releasing the data, but this still increases the wait time. Moreover, I suspect there might be a bug either in Next.js or tanstack-query, as such a problem does not exist in plain React.

I also noticed that Next.js processes server requests sequentially. For example, if two requests are sent to the Next.js server, they will be processed one after the other, not concurrently. This might be affecting the situation, although tanstack-query operates on the client side and should not be dependent on the Next.js server. I'll continue investigating.

**Reply by TkDodo** [reply]
> Here's the link [LINK](https://codesandbox.io/p/github/Sevansi/tanstack-query-test-react/draft/crazy-carlos?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clv9chz6d0006356iyl2los92%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clv9chz6d0002356i7xt10dmg%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clv9chz6d0004356i3v6sgbf4%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clv9chz6d0005356i9xeaslbw%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clv9chz6d0002356i7xt10dmg%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv9chz6c0001356iqntrwyu6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%255D%252C%2522id%2522%253A%2522clv9chz6d0002356i7xt10dmg%2522%252C%2522activeTabId%2522%253A%2522clv9chz6c0001356iqntrwyu6%2522%257D%252C%2522clv9chz6d0005356i9xeaslbw%2522%253A%257B%2522id%2522%253A%2522clv9chz6d0005356i9xeaslbw%2522%252C%2522activeTabId%2522%253A%2522clv9cj6960062356iwd9srd07%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clv9cj6960062356iwd9srd07%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clv9chz6d0004356i3v6sgbf4%2522%253A%257B%2522id%2522%253A%2522clv9chz6d0004356i3v6sgbf4%2522%252C%2522activeTabId%2522%253A%2522clv9cj08c004a356i08wd8l3r%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522id%2522%253A%2522clv9cj08c004a356i08wd8l3r%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Sorry, I can't access that. Seems to be a private sandbox.

> Moreover, I suspect there might be a bug either in Next.js or tanstack-query, as such a problem does not exist in plain React.

I think we should try to reproduce with `react-router` and `react-query`, and if that works, the issue seems to be on next-js side. If it also occurs in this setup, it' s on RQ's side.

**Reply by Sevansi** [reply]
I apologize, there was an error, and it became private, although it's set to public in the settings. I recreated the sandbox with React + tanstack-query, and there are no errors there! It seems that tanstack-query may have some compatibility issues with Next.js. I'll try other builds when I have more time. :)

[New Link](https://codesandbox.io/p/github/Sevansi/tanstack-query-test-react/main)

**Reply by Sevansi** [reply]
I can assume that the error might be on the side of tanstack-query. I tried many different versions and methods, and here's what I noticed: first, I installed Next.js 14.0.1 and tried it with react-query when it was still called that :) I got the same error. Then, I tried using window.history.pushState for a modal window, which seemed to work. Using it upon closing doesn't cause the page to hang, but in all other cases, I don't know why, it doesn't help. Then I tried changing the address bar in useEffect, and here window.history.pushState no longer helps. Most importantly, I tried placing two requests at page load for 10 seconds, and during them, I changed the address bar. Here, the behavior is very strange: the first request goes through despite the address bar change, but the second one does not; it hangs. I don’t even know where to dig or what to do next. :)

P.S. I also tried using window.history.pushState with the request that lasts 10 seconds, and it also hangs with it.

**Reply by Sevansi** [reply, marked as answer]
I found the problem, and on one hand, I'm glad, but on the other, I'm disappointed that I now need to rewrite 20 requests. :) The issue lies with `Next.js`, specifically in their new approach where you can use requests on the server side in a regular file using` 'use server'`—I forgot what this approach is called. The problem is that it breaks any requests if they happen simultaneously and there is a change in the address bar. I tried writing my function through `fetch `to the data retrieval function where `'use server' `is used and got the same behavior as with `tanstack-query`. The solution is very simple: use **`API routes`** for data retrieval. In this case, both tanstack-query and fetch, and everything else works fine, at least I haven't noticed any errors and bugs as with 'use server'. It would be advisable to report this issue to Next.js so that they can address it in future versions.
